### PR TITLE
Fix regression in regression

### LINKED ISSUE #11583: Wrong results for GLUE task STS-B
## Environment info
- `transformers` version: 4.6.0.dev0
- Platform: Linux-4.13.0-26-generic-x86_64-with-glibc2.10
- Python version: 3.8.8
- PyTorch version (GPU?): 1.8.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: True
- Using distributed or parallel set-up in script?: False

### Who can help
 @sgugger, @patil-suraj

## Information

Model I am using (Bert, XLNet ...): bert-base-cased

The problem arises when using:
* [ ] the official example scripts:  examples/pytorch/text-classification/run_glue.py

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: STS-B

## To reproduce

export TASK_NAME=stsb

python run_glue.py \
  --model_name_or_path bert-base-cased \
  --task_name $TASK_NAME \
  --do_train \
  --do_eval \
  --max_seq_length 128 \
  --per_device_train_batch_size 32 \
  --learning_rate 2e-5 \
  --num_train_epochs 3 \
  --output_dir /tmp/$TASK_NAME/

## Expected behavior
According to README, Person/Spearman corr. should be close to 88.64/88.48
However, I get Person/Spearman corr. equal to 28.36/27.70

The following warning occurs:
python3.8/site-packages/torch/nn/modules/loss.py:528: UserWarning: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.mse_loss(input, target, reduction=self.reduction)

### LINKED ISSUE #11780: Unintentional(?) interface change on loss function in models didn't work well for single-column regression
The recent PR #11012 changed the interface of forward function for `labels` in regression tasks as it skips `.view(-1)`  in loss function like [this](https://github.com/huggingface/transformers/pull/11012/files#diff-a48ba7f6444ca4954a58f1ac3e66c7941a2bbc4615649d56b182aeac8cc36d9cL1523). 
As shown below, that causes `UserWarning: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1]))` with BERT in glue example code, and it looks like this change was applied **not only to BERT but also a lot of models in #11012**

To resolve it, the current example glue code needs `if` statement that transform `labels` variable before `forward` function only for regression task.
But if the interface change in the PR was unintentional, I think we should revert `.view(-1)` in loss function.

## Environment info

- `transformers` version: 4.6.0
- Platform: Google Colab
- Python version: 3.7
- PyTorch version (GPU?): 1.8.1
- Tensorflow version (GPU?):
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help
@abhi1thakur @sgugger @LysandreJik from #11012

## Information

Model I am using (Bert, XLNet ...): `bert-base-uncased`

The problem arises when using:
* [x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [x] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

1. Run 
```
mkdir /tmp/stsb/ -p

python transformers/examples/pytorch/text-classification/run_glue_no_trainer.py \
  --model_name_or_path bert-base-cased \
  --task_name stsb \
  --max_length 128 \
  --per_device_train_batch_size 32 \
  --learning_rate 2e-5 \
  --num_train_epochs 3 \
  --output_dir /tmp/stsb/
```

2. We will see `UserWarning: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.` which didn't appear with the previous version of transformers like 2-3 weeks ago.  

```
05/19/2021 22:39:45 - INFO - __main__ -   ***** Running training *****
05/19/2021 22:39:45 - INFO - __main__ -     Num examples = 5749
05/19/2021 22:39:45 - INFO - __main__ -     Num Epochs = 3
05/19/2021 22:39:45 - INFO - __main__ -     Instantaneous batch size per device = 32
05/19/2021 22:39:45 - INFO - __main__ -     Total train batch size (w. parallel, distributed & accumulation) = 32
05/19/2021 22:39:45 - INFO - __main__ -     Gradient Accumulation steps = 1
05/19/2021 22:39:45 - INFO - __main__ -     Total optimization steps = 540
  0% 0/540 [00:00<?, ?it/s]/usr/local/lib/python3.7/dist-packages/torch/nn/modules/loss.py:528: UserWarning: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.mse_loss(input, target, reduction=self.reduction)
 33% 178/540 [00:24<00:47,  7.66it/s]/usr/local/lib/python3.7/dist-packages/torch/nn/modules/loss.py:528: UserWarning: Using a target size (torch.Size([21])) that is different to the input size (torch.Size([21, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.mse_loss(input, target, reduction=self.reduction)
 33% 180/540 [00:24<00:43,  8.26it/s]/usr/local/lib/python3.7/dist-packages/torch/nn/modules/loss.py:528: UserWarning: Using a target size (torch.Size([8])) that is different to the input size (torch.Size([8, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.mse_loss(input, target, reduction=self.reduction)
/usr/local/lib/python3.7/dist-packages/torch/nn/modules/loss.py:528: UserWarning: Using a target size (torch.Size([4])) that is different to the input size (torch.Size([4, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.mse_loss(input, target, reduction=self.reduction)
05/19/2021 22:40:13 - INFO - /usr/local/lib/python3.7/dist-packages/datasets/metric.py -   Removing /root/.cache/huggingface/metrics/glue/stsb/default_experiment-1-0.arrow
05/19/2021 22:40:13 - INFO - __main__ -   epoch 0: {'pearson': 0.40341441213742524, 'spearmanr': 0.41749739006146336}
 66% 359/540 [00:52<00:24,  7.26it/s]05/19/2021 22:40:40 - INFO - /usr/local/lib/python3.7/dist-packages/datasets/metric.py -   Removing /root/.cache/huggingface/metrics/glue/stsb/default_experiment-1-0.arrow
05/19/2021 22:40:40 - INFO - __main__ -   epoch 1: {'pearson': 0.4407148954008369, 'spearmanr': 0.4550002378117188}
100% 539/540 [01:20<00:00,  7.42it/s]05/19/2021 22:41:08 - INFO - /usr/local/lib/python3.7/dist-packages/datasets/metric.py -   Removing /root/.cache/huggingface/metrics/glue/stsb/default_experiment-1-0.arrow
05/19/2021 22:41:08 - INFO - __main__ -   epoch 2: {'pearson': 0.4408745967619131, 'spearmanr': 0.43830345183360847}
Configuration saved in /tmp/stsb/config.json
Model weights saved in /tmp/stsb/pytorch_model.bin
100% 540/540 [01:24<00:00,  6.42it/s]
```

As a result, this gave me a pretty bad performance `epoch 2: {'pearson': 0.4408745967619131, 'spearmanr': 0.43830345183360847}` while they were both around 0.87 with the previous version.

## Expected behavior
The following warning (actually should be a bug) should not appear, and the validation performance pearson and spearmanr should be around 0.87 with the parameters given in the example command.

`UserWarning: Using a target size (torch.Size([32])) that is different to the input size (torch.Size([32, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.`

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR fixes the regression introduced in #11012 for regression problems with only one label (like STS-B), see discussion on #11780. I checked both `run_glue` and `run_glue_no_trainer` on this branch and get the proper results for this task now.

Fixes #11780
Fixes #11583

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Yes, LGTM, thanks!

**Review by LysandreJik — APPROVED** [review]
Thanks for the test!

**Comment by yoshitomo-matsubara** [comment]
Thank you for fixing the issue!
